src/Network_Equivalences-Communication.thy

```diff
@@ -77,10 +108,544 @@ lemma inner_duplication_redundancy:
+          A \<guillemotleft> suffix n \<triangleright>\<^sup>\<infinity> x. \<Q> x \<guillemotleft> suffix n
+          \<rightarrow>\<^sub>s\<lparr>A \<guillemotleft> suffix n \<triangleright> \<star>\<^bsup>0\<^esup> X\<rparr>
+          (post_receive n X \<P> \<parallel> A \<guillemotleft> suffix n \<triangleright>\<^sup>\<infinity> x. \<P> x \<guillemotleft> suffix n) \<parallel>
+          post_receive 0 X (\<lambda>x. \<Q> x \<guillemotleft> suffix n \<parallel> A \<guillemotleft> suffix n \<triangleright>\<^sup>\<infinity> x. \<Q> x \<guillemotleft> suffix n)"
+          using synchronous_transition.parallel_right_io [where n = 0]
+          by (transfer, simp)
+        ultimately have "
+          ((\<zero> \<parallel> A \<guillemotleft> suffix n \<triangleleft> X \<parallel> \<zero>) \<parallel> \<currency>\<^sup>+ A \<guillemotleft> suffix n) \<parallel>
+          ((post_receive n X \<P> \<parallel> A \<guillemotleft> suffix n \<triangleright>\<^sup>\<infinity> x. \<P> x \<guillemotleft> suffix n) \<parallel>
+           A \<guillemotleft> suffix n \<triangleright>\<^sup>\<infinity> x. \<Q> x \<guillemotleft> suffix n)
+          \<rightarrow>\<^sub>s\<lparr>\<tau>\<rparr>
+          ((\<zero> \<parallel> \<zero> \<parallel> \<zero>) \<parallel> \<currency>\<^sup>+ A \<guillemotleft> suffix n) \<parallel>
+          ((post_receive n X \<P> \<parallel> A \<guillemotleft> suffix n \<triangleright>\<^sup>\<infinity> x. \<P> x \<guillemotleft> suffix n) \<parallel>
+           post_receive 0 X (\<lambda>x. \<Q> x \<guillemotleft> suffix n \<parallel> A \<guillemotleft> suffix n \<triangleright>\<^sup>\<infinity> x. \<Q> x \<guillemotleft> suffix n))"
+          using communication
+          by fastforce
+        then have "
+          ((\<zero> \<parallel> (A \<guillemotleft> suffix n) \<triangleleft> X \<parallel> \<zero>) \<parallel> \<currency>\<^sup>+ A \<guillemotleft> suffix n) \<parallel>
+          ((post_receive n X \<P> \<parallel> A \<guillemotleft> suffix n \<triangleright>\<^sup>\<infinity> x. \<P> x \<guillemotleft> suffix n) \<parallel>
+           A \<guillemotleft> suffix n \<triangleright>\<^sup>\<infinity> x. \<Q> x \<guillemotleft> suffix n)
+          \<rightarrow>\<^sub>s\<lparr>\<tau>\<rparr>
+          ((\<zero> \<parallel> \<zero> \<parallel> \<zero>) \<parallel> \<currency>\<^sup>+ A \<guillemotleft> suffix n) \<parallel>
+          ((post_receive n X \<P> \<parallel> A \<guillemotleft> suffix n \<triangleright>\<^sup>\<infinity> x. \<P> x \<guillemotleft> suffix n) \<parallel>
+           (post_receive n X \<Q> \<parallel> A \<guillemotleft> suffix n \<triangleright>\<^sup>\<infinity> x. \<Q> x \<guillemotleft> suffix n))"
+          unfolding post_receive_def
+          by (transfer, simp)
+        then show ?thesis .
+      qed
+      ultimately show ?thesis
+        by (simp, blast intro: rtranclp_trans)
+    qed
```

Comment:
The statement you prove here is a bit more complex than needed and, most importantly, its proof is very long, which makes it a bit intimidating and not so easy to follow. I came up with an alternative solution, in particular a greatly simplified proof.

My simplification relies on a few general lemmas, which we might want to add to the Þ-calculus:
```isabelle
lemma no_mobility_communication:
  assumes "η ≠ μ" and "P →⇩s⦇IO η A 0 X⦈ P'" and "Q →⇩s⦇IO μ A 0 X⦈ Q'"
  shows "P ∥ Q →⇩s⦇τ⦈ P' ∥ Q'"
  using communication [where n = 0] and assms
  unfolding funpow_0 .

lemma no_mobility_parallel_left_io:
  assumes "P →⇩s⦇IO η A 0 X⦈ P'"
  shows "P ∥ Q →⇩s⦇IO η A 0 X⦈ P' ∥ Q"
  using parallel_left_io [where n = 0] and assms
  by transfer simp

lemma no_mobility_parallel_right_io:
  assumes "Q →⇩s⦇IO η A 0 X⦈ Q'"
  shows "P ∥ Q →⇩s⦇IO η A 0 X⦈ P ∥ Q'"
  using parallel_right_io [where n = 0] and assms
  by transfer simp
```

Furthermore, my simplification introduces the following lemma, which we probably should keep private:
```isabelle
lemma adapted_repeated_receive_no_mobility_transition:
  shows "
    (A ▹⇧∞ x. 𝒫 x) « suffix n
    →⇩s⦇A « suffix n ▹ ⋆⇗0⇖ X⦈
    post_receive n X 𝒫 ∥ (A ▹⇧∞ x. 𝒫 x) « suffix n"
  unfolding adapted_after_repeated_receive
  using repeated_receive_transition [where 𝒫 = "λx. 𝒫 x « suffix n" and n = 0]
  unfolding post_receive_def
  by transfer force
```

With these preliminaries sorted out, the above code could be replaced by the following one:
```isabelle
    have "
      ¤⇧+ A ∥ A ▹⇧∞ x. 𝒫 x ∥ A ▹⇧∞ x. 𝒬 x
      ⇒⇩s⦇A ▹ ⋆⇗n⇖ X⦈
      ((𝟬 ∥ 𝟬 ∥ 𝟬) ∥ ¤⇧+ A « suffix n) ∥
      (post_receive n X 𝒫 ∥ (A ▹⇧∞ x. 𝒫 x) « suffix n) ∥
      (post_receive n X 𝒬 ∥ (A ▹⇧∞ x. 𝒬 x) « suffix n)"
    proof -
      have "
        ¤⇧+ A ∥ A ▹⇧∞ x. 𝒫 x ∥ A ▹⇧∞ x. 𝒬 x
        →⇩s⦇A ▹ ⋆⇗n⇖ X⦈
        ((A « suffix n ◃ X ∥ A « suffix n ◃ X ∥ 𝟬) ∥ ¤⇧+ A « suffix n) ∥
        (A ▹⇧∞ x. 𝒫 x) « suffix n ∥ (A ▹⇧∞ x. 𝒬 x) « suffix n"
        unfolding adapted_after_parallel [symmetric]
        by (intro duplication_transition parallel_left_io) 
      moreover have "… →⇩s⦇τ⦈
        ((𝟬 ∥ A « suffix n ◃ X ∥ 𝟬) ∥ ¤⇧+ A « suffix n) ∥
        (post_receive n X 𝒫 ∥ (A ▹⇧∞ x. 𝒫 x) « suffix n) ∥ (A ▹⇧∞ x. 𝒬 x) « suffix n"
        by
          (fast intro:
            no_mobility_communication
            no_mobility_parallel_left_io
            sending
            adapted_repeated_receive_no_mobility_transition
          )
      moreover have "… →⇩s⦇τ⦈
        ((𝟬 ∥ 𝟬 ∥ 𝟬) ∥ ¤⇧+ A « suffix n) ∥
        (post_receive n X 𝒫 ∥ (A ▹⇧∞ x. 𝒫 x) « suffix n) ∥
        (post_receive n X 𝒬 ∥ (A ▹⇧∞ x. 𝒬 x) « suffix n)"
        by
          (fast intro:
            no_mobility_communication
            no_mobility_parallel_left_io
            no_mobility_parallel_right_io
            sending
            adapted_repeated_receive_no_mobility_transition
          )
      ultimately show ?thesis
        by force
    qed
```

What do you think?

Reply:
Well, by employing `is`, the code can still be shrunk a bit: :grin:
```isabelle
    have "
      ¤⇧+ A ∥ A ▹⇧∞ x. 𝒫 x ∥ A ▹⇧∞ x. 𝒬 x
      ⇒⇩s⦇A ▹ ⋆⇗n⇖ X⦈
      ((𝟬 ∥ 𝟬 ∥ 𝟬) ∥ ¤⇧+ A « suffix n) ∥
      (post_receive n X 𝒫 ∥ (A ▹⇧∞ x. 𝒫 x) « suffix n) ∥
      (post_receive n X 𝒬 ∥ (A ▹⇧∞ x. 𝒬 x) « suffix n)"
      (is "?T ⇒⇩s⦇_⦈ ?U")
    proof -
      have "?T →⇩s⦇A ▹ ⋆⇗n⇖ X⦈
        ((A « suffix n ◃ X ∥ A « suffix n ◃ X ∥ 𝟬) ∥ ¤⇧+ A « suffix n) ∥
        (A ▹⇧∞ x. 𝒫 x) « suffix n ∥ (A ▹⇧∞ x. 𝒬 x) « suffix n"
        unfolding adapted_after_parallel [symmetric]
        by (intro duplication_transition parallel_left_io) 
      moreover have "… →⇩s⦇τ⦈
        ((𝟬 ∥ A « suffix n ◃ X ∥ 𝟬) ∥ ¤⇧+ A « suffix n) ∥
        (post_receive n X 𝒫 ∥ (A ▹⇧∞ x. 𝒫 x) « suffix n) ∥ (A ▹⇧∞ x. 𝒬 x) « suffix n"
        by
          (fast intro:
            no_mobility_communication
            no_mobility_parallel_left_io
            sending
            adapted_repeated_receive_no_mobility_transition
          )
      moreover have "… →⇩s⦇τ⦈ ?U"
        by
          (fast intro:
            no_mobility_communication
            no_mobility_parallel_left_io
            no_mobility_parallel_right_io
            sending
            adapted_repeated_receive_no_mobility_transition
          )
      ultimately show ?thesis
        by force
    qed
```

Reply:
I think, in other bisimilarity proofs, we were rather explicit when it came to constructing transitions. Your current proof above could be seen as following that style. However, I think what worked with very simple constructions does not work well with this more complex construction: while the reader is taken by the hand and led step by step through the construction, he might lose track, because it’s difficult to get a good overview of the construction due to its complexity.

Reply:
I think we should implement this simplification.

Reply:
However, although much shorter, I'm not yet so convinced that bundling the application of three transition rules such as `sending`, `parallel_left_io` and `communication` plus a pseudo-transition rule such as `adapted_repeated_receive_no_mobility_transition` into a single proof step actually improves readability. :thinking:.

Reply:
At least, we say in the proof method expression that we use `adapted_repeated_receive_no_mobility_transition`, as well as `sending` and `communication`. As a result, the reader can at least figure out with reasonable effort what’s going on. I’d say making a `sending` and a `receiving` transition and combining them with `communication` is a quite natural thing for someone used to process calculi. Therefore, the only real issue I see is the use of the non-standard `adapted_receive_no_mobility_transition`, but at least this has been introduced shortly before.

I think there are two competing effects at play here:

  * More detailed proofs make the entire construction explicit and this way aid understanding.
  * More concise proofs make the overall structure of the construction clearer and this way aid understanding.

I think the importance of the second point varies greatly with the complexity of the construction: while for simple constructions, like combining `receive` with `parallel_left_io`, it is of virtually no relevance, it is of considerable importance for constructions like the one we’re talking about here. Therefore, I’d say for simple constructions it is not so clear whether being concise is better or not, but it is rather clear that it is better for complex constructions.

Let’s go for this concise version.

Reply:
The no-mobility versions of transition system rules that the concise version uses have been introduced by input-output-hk/thorn-calculus#71 now.

Reply:
Please note that the PR mentioned above has already been accepted.